ui/discovery/ssdp.go

```diff
@@ -0,0 +1,181 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package discovery
+
+import (
+	"bytes"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"runtime"
+	"sync"
+	"text/template"
+	"time"
+
+	log "github.com/cihub/seelog"
+	"github.com/gofrs/uuid"
+	"github.com/koron/go-ssdp"
+	"github.com/mysteriumnetwork/node/metadata"
+	"github.com/pkg/errors"
+)
+
+const ssdpLogPrefix = "[SSDP] "
+
+const deviceDescriptionTemplate = `<?xml version="1.0"?>
+<root xmlns="urn:schemas-upnp-org:device-1-0" configId="1">
+	<specVersion>
+		<major>1</major>
+		<minor>1</minor>
+	</specVersion>
+	<device>
+		<deviceType>urn:schemas-upnp-org:device:node:1</deviceType>
+		<friendlyName>Mysterium Node</friendlyName>
+
+		<manufacturer>Mysterium Network</manufacturer>
+		<manufacturerURL>https://mysterium.network/</manufacturerURL>
+
+		<modelName>Raspberry Node</modelName>
+		<modelNumber>{{.Version}}</modelNumber>
+		<modelURL>https://mysterium.network/node/</modelURL>
+
+		<serialNumber>1234</serialNumber>
+		<UDN>uuid:{{.UUID}}</UDN>
+		<presentationURL>{{.URL}}</presentationURL>
+	</device>
+</root>`
+
+type ssdpServer struct {
+	url  string
+	uuid string
+	ssdp *ssdp.Advertiser
+	quit chan struct{}
+	once sync.Once
+}
+
+func newSSDPServer(url string) *ssdpServer {
+	return &ssdpServer{
+		url:  url,
+		quit: make(chan struct{}),
+	}
+}
+
+func (ss *ssdpServer) Start() (err error) {
+	ss.uuid, err = generateUUID()
+	if err != nil {
+		return errors.Wrap(err, "failed to generate new UUID")
+	}
+
+	url, err := ss.serveDeviceDescriptionDocument()
+	if err != nil {
+		return errors.Wrap(err, "failed to serve device description document")
+	}
+
+	ss.ssdp, err = ssdp.Advertise(
+		"upnp:rootdevice",                   // ST: Type
+		"uuid:"+ss.uuid+"::upnp:rootdevice", // USN: ID
+		url.String(),                        // LOCATION: location header
+		runtime.GOOS+" UPnP/1.1 node/"+metadata.VersionAsString(), // SERVER: server header
+		1800) // cache control, max-age
+	if err != nil {
+		return errors.Wrap(err, "failed to start SSDP advertiser")
+	}
+
+	for {
+		select {
+		case <-time.After(30 * time.Second):
+			if err := ss.ssdp.Alive(); err != nil {
+				log.Warn(ssdpLogPrefix, "failed to sent SSDP Alive message: ", err)
+			}
+		case <-ss.quit:
+			return nil
+		}
+	}
+}
+
+func (ss *ssdpServer) Stop() error {
+	ss.once.Do(func() {
+		close(ss.quit)
+	})
+
+	if err := ss.ssdp.Bye(); err != nil {
```

Comment:
`ss.quit` is just an announcement to stop async processes. 
`ss.ssdp` server is alive when we are sending `Bye`. It will be closed with a `ss.ssdp.Close()` lower.